src/System.Linq.Queryable/tests/QueryFromExpressionTests.cs

```diff
@@ -0,0 +1,508 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+using System.Reflection;
+using Xunit;
+
+namespace System.Linq.Tests
+{
+    public class QueryFromExpressionTests
+    {
+        private class SimplePair : IEnumerable<int>
+        {
+            public int First { get; set; }
+
+            public int Second { get; set; }
+
+            public IEnumerator<int> GetEnumerator()
+            {
+                yield return First;
+                yield return Second;
+            }
+
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return GetEnumerator();
+            }
+        }
+
+        private static IQueryProvider _prov = Enumerable.Empty<int>().AsQueryable().Provider;
+
+        [Fact]
+        public void ConditionalEqualModuloConstantConstant()
+        {
+            Expression cond = Expression.Condition(
+                Expression.Equal(
+                    Expression.Modulo(Expression.Constant(1), Expression.Constant(2)),
+                    Expression.Constant(0)
+                   ),
+                Expression.Constant(Enumerable.Range(0, 2).AsQueryable()),
+                Expression.Constant(Enumerable.Range(3, 2).AsQueryable())
+            );
+            IQueryable<int> q = _prov.CreateQuery<int>(cond);
+            Assert.Equal(Enumerable.Range(3, 2), q);
+        }
+
+        [Fact]
+        public void ConditionalEqualModuloConstantConstantExplicitConstruction()
+        {
+            Expression cond = Expression.Condition(
+                Expression.Equal(
+                    Expression.Modulo(Expression.Constant(1), Expression.Constant(2)),
+                    Expression.Constant(0)
+                   ),
+                Expression.Constant(Enumerable.Range(0, 2).AsQueryable()),
+                Expression.Constant(Enumerable.Range(3, 2).AsQueryable())
+            );
+            IQueryable<int> q = new EnumerableQuery<int>(cond);
+            Assert.Equal(Enumerable.Range(3, 2), q);
+        }
+
+        private static IQueryable<char> AsciiControlCharacters
+        {
+            get
+            {
+                return Enumerable.Range(0, 128).AsQueryable()
+                    .Select(i => (char)i)
+                    .Where(c => char.IsControl(c));
+            }
+        }
+
+        [Fact]
+        public void PropertyAccess()
+        {
+            Expression access = Expression.Property(null, typeof(QueryFromExpressionTests), "AsciiControlCharacters");
+            IQueryable<char> q = _prov.CreateQuery<char>(access);
+            Assert.Equal(Enumerable.Range(0, 32).Concat(Enumerable.Range(127, 1)).Select(i => (char)i), q);
```

Comment:
Would it not be cleaner to just compare against the same ```Enumerable.Range(0, 128).Select(i => (char)i).Where(c => char.IsControl(c))``` so that we can easily see that we expect as output what was given as input?